core/src/main/java/org/apache/iceberg/ManifestWriter.java

```diff
@@ -139,13 +141,33 @@ void add(ManifestEntry<F> entry) {
    * @param existingFile a file
    * @param fileSnapshotId snapshot ID when the data file was added to the table
    * @param dataSequenceNumber a data sequence number of the file (assigned when the file was added)
+   * @deprecated since 1.1.0, will be removed in 1.2.0; use {@link #existing(ContentFile, long,
+   *     long, Long)}.
    */
+  @Deprecated
   public void existing(F existingFile, long fileSnapshotId, long dataSequenceNumber) {
-    addEntry(reused.wrapExisting(fileSnapshotId, dataSequenceNumber, existingFile));
+    addEntry(reused.wrapExisting(fileSnapshotId, dataSequenceNumber, null, existingFile));
+  }
+
+  /**
+   * Add an existing entry for a file.
+   *
+   * <p>The original data and file sequence numbers, snapshot ID, which were assigned at commit,
+   * must be preserved when adding an existing entry.
+   *
+   * @param existingFile a file
+   * @param fileSnapshotId snapshot ID when the data file was added to the table
+   * @param dataSequenceNumber a data sequence number of the file (assigned when the file was added)
+   * @param fileSequenceNumber a file sequence number (assigned when the file was added)
+   */
+  public void existing(
+      F existingFile, long fileSnapshotId, long dataSequenceNumber, Long fileSequenceNumber) {
```

Comment:
Note `fileSequenceNumber` is Long because the value may be null in old v2 manifests.